sdk/plugin/grpc_backend_client.go

```diff
@@ -278,3 +314,22 @@ func (b *backendGRPCPluginClient) Type() logical.BackendType {
 
 	return logical.BackendType(reply.Type)
 }
+
+// encodeNamespace embeds namespace data from the provided context into GRPC metadata.
+func encodeNamespace(ctx context.Context) (context.Context, error) {
+	if ctx == nil {
+		return ctx, nil
+	}
+
+	// Add namespace information, if present.
+	ns, err := namespace.FromContext(ctx)
+	if err == namespace.ErrNoNamespace {
+		return ctx, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	ctx = metadata.AppendToOutgoingContext(ctx, metadataNSID, ns.ID, metadataNSPath, ns.Path)
```

Comment:
I wonder if we should use `metadata.NewOutgoingContext` in here. I am not sure if this would panic if there's no metadata context value present (more specifically, not sure how [this `copy` call](https://github.com/hashicorp/vault/blob/7633a6fb7fcd34cbffa9d949b427f2a1edd3c34f/vendor/google.golang.org/grpc/metadata/metadata.go#L160) behaves if `md` is nil).

Reply:
Quick [playground demo](https://play.golang.org/p/WEgQbIhPN2A) indicates that it might..?

Reply:
Since the type assertion is to `rawMD` (not `*rawMD`) the result is the zero value even if the value being asserted is nil. Note the `_` ok term. I actually had to check that in the Go spec on that bit:

> The value of ok is true if the assertion holds. Otherwise it is false and the value of v is the zero value for type T. No run-time panic occurs in this case. 

So with a zero value `md`, you end up with a nil (but copyable) slice.  https://play.golang.org/p/F2DjM-e-skr

Reply:
Ahh, I good catch!